pkg/sql/conn_executor.go

```diff
@@ -1846,7 +1846,10 @@ func (ex *connExecutor) execCmd() error {
 			// in the batch. This matches the Postgres behavior. See
 			// "Multiple Statements in a Single Query" at
 			// https://www.postgresql.org/docs/14/protocol-flow.html.
-			canAutoCommit := ex.implicitTxn() && tcmd.LastInBatch
+			// The behavior is configurable, in case users want to preserve the
+			// behavior from v21.2 and earlier.
+			implicitTxnForBatch := ex.sessionData().EnableImplicitTransactionForBatchStatements
```

Comment:
wondering: if someone does `ALTER DATABASE d SET enable_implicit_transaction_for_batch_statements = false` and rolls back, does their database not load because the session variable cannot be set? i wonder if people would want this behaviour with mixed versions.

also, do we want to metamorphic test this or nah?

Reply:
> does their database not load because the session variable cannot be set

do you mean the setting that happens during session init? we should be good there. a warning is logged and it continues: https://github.com/cockroachdb/cockroach/blob/29797de1825eb1a12c357f899ef2308a60ebf8e0/pkg/sql/pgwire/auth.go#L189-L191

hm yeah i suppose metamorphic won't hurt

Reply:
actually, a metamorphic setting might hurt, since it would make traces non-deterministic. i guess we can try then see if tests are affected

Reply:
if the tests are tricky im happy to let it go. but i think this looks fine.

my only other concern here is you need to migrate at least one node to move before setting the default var. but then again, you could argue that users should have these implicit statements split out like you have in the tests in v21.2 before migrating to v22.1 anyway, so a backwards compatible pathway is available.